File: from-books/recursion/language/go/exercise-2.go
package main

import (
	"fmt"
)

func main() {
	mixed := []interface{}{
		1, 2, 3,
		[]int{4, 5, 6},
		7, 8,
		[]interface{}{
			9,
			[]int{10, 11},
		},
	}

	printNumbers(mixed)
}

// Create a function that prints all the numbers in a mixed array with multiple layers.

func printNumbers(data interface{}) {
	switch v := data.(type) {
	case int:
		fmt.Println(v)
	case []interface{}:
		for _, item := range v {
			printNumbers(item)
		}
	case []int:
		for _, item := range v {
			printNumbers(item)
		}
	}
}

File: from-books/recursion/language/rust/exercise-2.rs
enum NestedNumber {
    Num(i32),
    Nested(Vec<NestedNumber>),
}

fn print_numbers(data: &NestedNumber) {
    match data {
        NestedNumber::Num(x) => println!("{}", x),
        NestedNumber::Nested(vec) => {
            for item in vec {
                print_numbers(item);
            }
        }
    }
}

fn main() {
    let mixed = NestedNumber::Nested(vec![
        NestedNumber::Num(1),
        NestedNumber::Num(2),
        NestedNumber::Num(3),
        NestedNumber::Nested(vec![
            NestedNumber::Num(4),
            NestedNumber::Num(5),
            NestedNumber::Num(6),
        ]),
        NestedNumber::Num(7),
        NestedNumber::Num(8),
        NestedNumber::Nested(vec![
            NestedNumber::Num(9),
            NestedNumber::Nested(vec![NestedNumber::Num(10), NestedNumber::Num(11)]),
        ]),
    ]);

    print_numbers(&mixed);
}
